Remove undefined behavior from shifts

ISSOtm opened the pull request:
Fixes #431.
`asl` and `asr` functions in `src/link/patch.c` courtesy of @PinoBatch.
(Reorganized that code in RGBASM evaluators to sprinkle warnings)

AntonioND commented on src/asm/constexpr.c:
```diff
@@ -209,28 +209,53 @@ void constexpr_BinaryOp(struct ConstExpression *expr,
+				warning(WARNING_SHIFT_AMOUNT, "Shifting right by negative amount %d",
+					value2);
+			value2 = -value2; // Treat right shift as negative left
+shift:
```
Let's avoid gotos... I think this should be a function.

ISSOtm replied:
I agree, but it seemed too specific at the time. Such a simple `goto` seemed like it wouldn't hurt, and also preserves locality/context.

aaaaaa123456789 replied:
Very much harm has been done by the widespread myth that `goto` is inherently wrong. People might argue that non-linear flow is complicated, but every single flow control keyword introduces non-linear code flow and that has (fortunately) never stopped anyone from using them.

The reason why people spoke loudly against `goto` half a century ago was because, with many programmers being used to assembly and to the idea of creating control structures by jumping around, people used `goto` instead of the proper flow control keywords that higher-level languages introduced. This remains a valid complaint, but _nobody does that nowadays_. There's no reason to avoid using `goto` when you really do want to have multiple code paths concur at a point; it's the natural way to do it.

All in all, while people have been historically right when complaining about others using `goto` to reinvent the `while` loop, that's not what people use it for nowadays, and therefore there's no reason to treat it as an evil thing that should not be touched.

AntonioND replied:
IN general, ``goto`` is the wrong thing to do. It has some nice uses like returning from a function with several error conditions that also needs to allocate resources:
```
int main(void)
{
    int ret;

    void *p = malloc(100);

    if (function1() == ERROR) {
        ret = -1;
        goto cleanup;
    }

    if (function2() == ERROR) {
        ret = -2;
        goto cleanup;
    }
        
    ret = 0;
cleanup:
    free(p);
    return ret;
}
```
In other situation, if you are forced to use ``goto``, it is because the code flow is a mess. If not, for the most part, you don't need it.

aaaaaa123456789 commented:
As I've said in other places, the warning for shifting negative values seems like a bad idea. This is a very well-defined operation with a very well-defined result, and having to litter code with often-hard-to-calculate bit-clearing constants doesn't help anyone.  
(Anyone who thinks that `((\1) & ((1 << (-(\2) & 31)) - 1)) << (\2)` is clearer than `(\1) << (\2)` feel free to correct me; I had to stop to count the parens in that first expression like four times before getting them right.)

ISSOtm commented:
Removed the `goto`, though I like this solution less because it duplicates the `shr` check.
Removed the warning when left shifting negative values. That said, again, you could just have disabled the warning if you didn't care about it.
